jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java

```diff
@@ -0,0 +1,193 @@
+    Future<Optional<String>> messageFuture =
+        executorService.submit(
+            () ->
+                performUpdateCheck(
+                    Verify.verifyNotNull(ProjectInfo.VERSION), versionUrl, getConfigDir()));
+    executorService.shutdown();
+    return new UpdateChecker(messageFuture);
+  }
+
+  @VisibleForTesting
+  static Optional<String> performUpdateCheck(
+      String currentVersion, String versionUrl, Path configDir) {
+    // Abort if offline or update checks are disabled
+    if (Boolean.getBoolean(PropertyNames.DISABLE_UPDATE_CHECKS)) {
+      return Optional.empty();
+    }
+
+    Path configFile = configDir.resolve("config.json");
+    Path lastUpdateCheck = configDir.resolve("lastUpdateCheck");
+    try {
+      if (Files.exists(configFile)) {
+        // Abort if update checks are disabled
+        ConfigJsonTemplate config =
+            JsonTemplateMapper.readJsonFromFile(configFile, ConfigJsonTemplate.class);
+        if (config.disableUpdateCheck) {
+          return Optional.empty();
+        }
+      } else {
+        // Generate config file if it doesn't exist
+        ConfigJsonTemplate config = new ConfigJsonTemplate();
+        Files.createDirectories(configDir);
+        try (OutputStream outputStream = Files.newOutputStream(configFile)) {
+          JsonTemplateMapper.writeTo(config, outputStream);
+        }
+      }
+
+      // Check time of last update check
+      if (Files.exists(lastUpdateCheck)) {
+        String fileContents =
+            new String(Files.readAllBytes(lastUpdateCheck), StandardCharsets.UTF_8);
+        try {
+          Instant modifiedTime = Instant.parse(fileContents);
+          if (modifiedTime.plus(Duration.ofDays(1)).isAfter(Instant.now())) {
+            return Optional.empty();
+          }
+        } catch (DateTimeParseException ex) {
+          // Ignore parse failure; assume update check hasn't been performed
+        }
+      }
+
+      // Check for update
+      HttpURLConnection connection = (HttpURLConnection) new URL(versionUrl).openConnection();
+      try {
+        connection.setConnectTimeout(3000);
+        BufferedReader bufferedReader =
+            new BufferedReader(
+                new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8));
+        String latestVersion = bufferedReader.readLine().trim();
+        Files.write(lastUpdateCheck, Instant.now().toString().getBytes(StandardCharsets.UTF_8));
+        if (currentVersion.equals(latestVersion)) {
+          return Optional.empty();
+        }
+        return Optional.of(
+            "A new version of Jib ("
+                + latestVersion
+                + ") is available (currently using "
+                + currentVersion
+                + "). Update your build configuration to use the latest features and fixes!");
+
+      } finally {
+        connection.disconnect();
+      }
+
+    } catch (IOException ex) {
+      try {
+        Files.deleteIfExists(configFile);
```

Comment:
I'm not sure I understand what's going on here? why do we delete these files?

Reply:
Was suggested by @chanseokoh to avoid corruption in case the thread is interrupted while writing to a file.

Reply:
About corruption: theoretically the files are shared resources among multiple concurrent Jib runs, although the likelihood is very low that the files are concurrently modified. That's why I am OK with not doing any synchronization for a shared object here.

I'm not saying that this would be the only source of possible corruption. But if the files were ever corrupted somehow, Jib would not be able to recover (it will just throw an `IOExcetion` again and again) without deleting the corrupted files. I just want it to be able to recover.

Reply:
And yeah, I think thread interrupted might be another source of possible corruption.

Reply:
deleting the config file could potentially result in the user's non-update-check config to be deleted no?

Reply:
I don't think we should ever delete the user's config file.

Reply:
We can output a message saying your config is corrupt? And to go clean it up?

Reply:
Yeah, I wanted a simple solution, but in principle I agree it's better to let the user delete the files.

Reply:
Also, if it is not corruption but a temporary IO error, I think it sill makes sense to report the error saying that Jib couldn't read its config file (but not a fatal error).